Not reconnecting when MQTT broker not available

I have MQTT-IO on a different system from where MQTT is installed.
If I have to reboot the raspberry pi where MQTT is running, MQTT-IO is loosing connection but not trying to connect anymore.

When MQTT broker come up, any event MQTT-IO tries to publish is reporting the following error:

```
2021-03-15 08:07:40 mqtt_io.server [INFO] Digital input 'alarm_out1' value changed to False
2021-03-15 08:07:40 mqtt_io.server [ERROR] Exception while handling MQTT task:
Traceback (most recent call last):
  File "/home/pi/.local/lib/python3.7/site-packages/mqtt_io/server.py", line 996, in _mqtt_task_loop await entry.coro
  File "/home/pi/.local/lib/python3.7/site-packages/mqtt_io/server.py", line 635, in _mqtt_publish await self.mqtt.publish(msg)
  File "/home/pi/.local/lib/python3.7/site-packages/mqtt_io/mqtt/asyncio_mqtt.py", line 76, in publish topic=msg.topic, payload=msg.payload, qos=msg.qos, retain=msg.retain
  File "/usr/local/lib/python3.7/dist-packages/asyncio_mqtt/client.py", line 128, in publish raise MqttCodeError(info.rc, 'Could not publish message')
asyncio_mqtt.error.MqttCodeError: [code:4] Could not publish message
2021-03-15 08:14:04 mqtt_io.server [INFO] Waiting for our tasks to complete...
Traceback (most recent call last):
  File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main "__main__", mod_spec)
  File "/usr/lib/python3.7/runpy.py", line 85, in _run_code  exec(code, run_globals)
  File "/home/pi/.local/lib/python3.7/site-packages/mqtt_io/__main__.py", line 38, in <module> main()
  File "/home/pi/.local/lib/python3.7/site-packages/mqtt_io/__main__.py", line 33, in main mqtt_gpio.run()
  File "/home/pi/.local/lib/python3.7/site-packages/mqtt_io/server.py", line 1174, in run self.loop.run_until_complete(self._main_task)
  File "/usr/lib/python3.7/asyncio/base_events.py", line 584, in run_until_complete return future.result()
  File "/home/pi/.local/lib/python3.7/site-packages/mqtt_io/server.py", line 1168, in run await self.shutdown()
  File "/home/pi/.local/lib/python3.7/site-packages/mqtt_io/server.py", line 1229, in shutdown wait=False,
  File "/home/pi/.local/lib/python3.7/site-packages/mqtt_io/server.py", line 635, in _mqtt_publish await self.mqtt.publish(msg)
  File "/home/pi/.local/lib/python3.7/site-packages/mqtt_io/mqtt/asyncio_mqtt.py", line 76, in publish topic=msg.topic, payload=msg.payload, qos=msg.qos, retain=msg.retain
  File "/usr/local/lib/python3.7/dist-packages/asyncio_mqtt/client.py", line 128, in publish raise MqttCodeError(info.rc, 'Could not publish message')
asyncio_mqtt.error.MqttCodeError: [code:4] Could not publish message
```

--- a/mqtt_io/config/config.schema.yml
+++ b/mqtt_io/config/config.schema.yml
@@ -250,7 +250,24 @@ mqtt:
           type: boolean
           required: no
           default: false
-
+    reconnect-delay:
+      meta:
+        description: |
+          Time in seconds to wait between reconnect attempts.
+      type: integer
+      required: no
+      default: 2
+      min: 1
+    reconnect-count:
+      meta:
+        description: |
+          Max number of retries of connections before giving up (and exiting).
+          -1 means infinite reconnects.
+          The counter is reset if a connection was established successfully.
+      type: integer
+      required: no
+      default: 0
+      min: -1
 gpio_modules:
   meta:
     description: |

--- a/mqtt_io/config/config.schema.yml
+++ b/mqtt_io/config/config.schema.yml
@@ -250,7 +250,7 @@ mqtt:
           type: boolean
           required: no
           default: false
-    reconnect-delay:
+    reconnect_delay:
       meta:
         description: |
           Time in seconds to wait between reconnect attempts.
@@ -258,7 +258,7 @@ mqtt:
       required: no
       default: 2
       min: 1
-    reconnect-count:
+    reconnect_count:
       meta:
         description: |
           Max number of retries of connections before giving up (and exiting).

--- a/mqtt_io/config/config.schema.yml
+++ b/mqtt_io/config/config.schema.yml
@@ -262,11 +262,11 @@ mqtt:
       meta:
         description: |
           Max number of retries of connections before giving up (and exiting).
-          -1 means infinite reconnects.
+          -1 (the default) means infinite reconnects.
           The counter is reset if a connection was established successfully.
       type: integer
       required: no
-      default: 0
+      default: -1
       min: -1
 gpio_modules:
   meta:

--- a/mqtt_io/mqtt/__init__.py
+++ b/mqtt_io/mqtt/__init__.py
@@ -162,3 +162,7 @@ class AbstractMQTTClient(abc.ABC):
         client: Type[AbstractMQTTClient]
         client = client_module.MQTTClient  # type: ignore[attr-defined]
         return client
+
+
+class MQTTException(Exception):
+    pass

--- a/mqtt_io/mqtt/__init__.py
+++ b/mqtt_io/mqtt/__init__.py
@@ -165,4 +165,6 @@ class AbstractMQTTClient(abc.ABC):
 
 
 class MQTTException(Exception):
-    pass
+    """
+    An Exception which should be raised on errors in implementations of AbstractMQTTClient
+    """

--- a/mqtt_io/mqtt/asyncio_mqtt.py
+++ b/mqtt_io/mqtt/asyncio_mqtt.py
@@ -8,7 +8,7 @@ from asyncio.queues import QueueFull
 from functools import wraps
 from typing import Any, List, Optional, Tuple, TypeVar, Callable, cast
 
-from asyncio_mqtt.client import Client, Will  # type: ignore
+from asyncio_mqtt.client import Client, Will, MqttError  # type: ignore
 from paho.mqtt import client as paho  # type: ignore
 
 from . import (
@@ -29,7 +29,7 @@ def _map_exception(func: Func) -> Func:
     async def inner(*args: Any, **kwargs: Any) -> Any:
         try:
             await func(*args, **kwargs)
-        except Exception as exc:
+        except MqttError as exc:
             raise MQTTException from exc
 
     return cast(Func, inner)

--- a/mqtt_io/mqtt/asyncio_mqtt.py
+++ b/mqtt_io/mqtt/asyncio_mqtt.py
@@ -27,8 +27,8 @@ def _map_exception(func):
     async def inner(*args, **kwargs):
         try:
             await func(*args, **kwargs)
-        except Exception as e:
-            raise MQTTException from e
+        except Exception as exc:
+            raise MQTTException from exc
 
     return inner
 

--- a/mqtt_io/mqtt/asyncio_mqtt.py
+++ b/mqtt_io/mqtt/asyncio_mqtt.py
@@ -5,21 +5,32 @@ Implementation of AbstractMQTTClient using asyncio-mqtt client.
 import asyncio
 import logging
 from asyncio.queues import QueueFull
+from functools import wraps
 from typing import Any, List, Optional, Tuple
 
-from paho.mqtt import client as paho  # type: ignore
-
 from asyncio_mqtt.client import Client, Will  # type: ignore
+from paho.mqtt import client as paho  # type: ignore
 
 from . import (
     AbstractMQTTClient,
     MQTTClientOptions,
     MQTTMessage,
     MQTTMessageSend,
-    MQTTProtocol,
+    MQTTProtocol, MQTTException,
 )
 
 _LOG = logging.getLogger(__name__)
+
+
+def _map_exception(func):
+    @wraps(func)
+    async def inner(*args, **kwargs):
+        try:
+            await func(*args, **kwargs)
+        except Exception as e:
+            raise MQTTException from e
+
+    return inner
 
 
 class MQTTClient(AbstractMQTTClient):
@@ -59,18 +70,22 @@ class MQTTClient(AbstractMQTTClient):
         )
         self._message_queue: Optional[asyncio.Queue[MQTTMessage]] = None
 
+    @_map_exception
     async def connect(self, timeout: int = 10) -> None:
         await self._client.connect(timeout=timeout)
 
+    @_map_exception
     async def disconnect(self) -> None:
         try:
             await self._client.disconnect()
         except TimeoutError:
             await self._client.force_disconnect()
 
+    @_map_exception
     async def subscribe(self, topics: List[Tuple[str, int]]) -> None:
         await self._client.subscribe(topics)
 
+    @_map_exception
     async def publish(self, msg: MQTTMessageSend) -> None:
         await self._client.publish(
             topic=msg.topic, payload=msg.payload, qos=msg.qos, retain=msg.retain

--- a/mqtt_io/mqtt/asyncio_mqtt.py
+++ b/mqtt_io/mqtt/asyncio_mqtt.py
@@ -6,7 +6,7 @@ import asyncio
 import logging
 from asyncio.queues import QueueFull
 from functools import wraps
-from typing import Any, List, Optional, Tuple
+from typing import Any, List, Optional, Tuple, TypeVar, Callable, cast
 
 from asyncio_mqtt.client import Client, Will  # type: ignore
 from paho.mqtt import client as paho  # type: ignore
@@ -21,16 +21,18 @@ from . import (
 
 _LOG = logging.getLogger(__name__)
 
+Func = TypeVar('Func', bound=Callable[..., Any])
 
-def _map_exception(func):
+
+def _map_exception(func: Func) -> Func:
     @wraps(func)
-    async def inner(*args, **kwargs):
+    async def inner(*args: Any, **kwargs: Any) -> Any:
         try:
             await func(*args, **kwargs)
         except Exception as exc:
             raise MQTTException from exc
 
-    return inner
+    return cast(Func, inner)
 
 
 class MQTTClient(AbstractMQTTClient):

--- a/mqtt_io/server.py
+++ b/mqtt_io/server.py
@@ -1184,6 +1184,8 @@ class MqttIo:  # pylint: disable=too-many-instance-attributes
             finally:
                 self.running.clear()
                 self.mqtt_connected.clear()
+                if self.critical_tasks:
+                    asyncio.gather(*self.critical_tasks, return_exceptions=True).cancel()
             if reconnect:
                 await asyncio.sleep(self.config['mqtt'].get('reconnect_delay'))
         await self.shutdown()

--- a/mqtt_io/server.py
+++ b/mqtt_io/server.py
@@ -19,7 +19,6 @@ from hashlib import sha1
 from importlib import import_module
 from typing import Any, Dict, List, Optional, Tuple, Type, Union, overload
 
-import asyncio_mqtt  # type: ignore
 import backoff  # type: ignore
 from typing_extensions import Literal
 
@@ -67,7 +66,7 @@ from .mqtt import (
     MQTTClientOptions,
     MQTTMessageSend,
     MQTTTLSOptions,
-    MQTTWill,
+    MQTTWill, MQTTException,
 )
 from .types import ConfigType, PinType, SensorValueType
 from .utils import PriorityCoro, create_unawaited_task_threadsafe
@@ -1175,7 +1174,7 @@ class MqttIo:  # pylint: disable=too-many-instance-attributes
                     _LOG.exception("Exception in critical task:")
             except asyncio.CancelledError:
                 break
-            except asyncio_mqtt.error.MqttError:
+            except MQTTException:
                 if counter > 0:
                     counter -= 1
                 _LOG.exception('Connection to MQTT-Broker failed')

--- a/mqtt_io/server.py
+++ b/mqtt_io/server.py
@@ -1175,11 +1175,11 @@ class MqttIo:  # pylint: disable=too-many-instance-attributes
             except asyncio.CancelledError:
                 break
             except MQTTException:
+                if counter == 0:
+                    break
                 if counter > 0:
                     counter -= 1
                 _LOG.exception('Connection to MQTT-Broker failed')
-                if counter == 0:
-                    break
                 await asyncio.sleep(self.config['mqtt'].get('reconnect_delay'))
             finally:
                 self.running.clear()

--- a/mqtt_io/server.py
+++ b/mqtt_io/server.py
@@ -1146,6 +1146,46 @@ class MqttIo:  # pylint: disable=too-many-instance-attributes
             else:
                 self.event_bus.fire(StreamDataSentEvent(stream_conf["name"], data))
 
+    async def _main_loop(self) -> None:
+        counter = self.config['mqtt'].get('reconnect-count')
+        while True:
+            try:
+                await self._connect_mqtt()
+                # Reset counter
+                counter = self.config['mqtt'].get('reconnect-count')
+                self.critical_tasks = [
+                    self.loop.create_task(coro)
+                    for coro in (
+                        self._mqtt_task_loop(),
+                        self._mqtt_rx_loop(),
+                        self._remove_finished_transient_tasks(),
+                    )
+                ]
+
+                self.running.set()
+
+                if self.config["mqtt"].get("ha_discovery", {}).get("enabled"):
+                    self._ha_discovery_announce()
+
+                try:
+                    await asyncio.gather(*self.critical_tasks)
+                except asyncio.CancelledError:
+                    break
+                except Exception:  # pylint: disable=broad-except
+                    _LOG.exception("Exception in critical task:")
+            except asyncio.CancelledError:
+                break
+            except asyncio_mqtt.error.MqttError:
+                if counter > 0:
+                    counter -= 1
+                _LOG.exception('Connection to MQTT-Broker failed')
+                if counter == 0:
+                    break
+                await asyncio.sleep(self.config['mqtt'].get('reconnect-delay'))
+            finally:
+                self.running.clear()
+        await self.shutdown()
+
     # Main entry point
 
     def run(self) -> None:
@@ -1171,47 +1211,7 @@ class MqttIo:  # pylint: disable=too-many-instance-attributes
         self._init_sensor_inputs()
         self._init_stream_modules()
 
-        async def run() -> None:
-            counter = self.config['mqtt'].get('reconnect-count')
-            while True:
-                try:
-                    await self._connect_mqtt()
-                    # Reset counter
-                    counter = self.config['mqtt'].get('reconnect-count')
-                    self.critical_tasks = [
-                        self.loop.create_task(coro)
-                        for coro in (
-                            self._mqtt_task_loop(),
-                            self._mqtt_rx_loop(),
-                            self._remove_finished_transient_tasks(),
-                        )
-                    ]
-
-                    self.running.set()
-
-                    if self.config["mqtt"].get("ha_discovery", {}).get("enabled"):
-                        self._ha_discovery_announce()
-
-                    try:
-                        await asyncio.gather(*self.critical_tasks)
-                    except asyncio.CancelledError:
-                        break
-                    except Exception:  # pylint: disable=broad-except
-                        _LOG.exception("Exception in critical task:")
-                except asyncio.CancelledError:
-                    break
-                except asyncio_mqtt.error.MqttError:
-                    if counter > 0:
-                        counter -= 1
-                    _LOG.exception('Connection to MQTT-Broker failed')
-                    if counter == 0:
-                        break
-                    await asyncio.sleep(self.config['mqtt'].get('reconnect-delay'))
-                finally:
-                    self.running.clear()
-            await self.shutdown()
-
-        self._main_task = self.loop.create_task(run())
+        self._main_task = self.loop.create_task(self._main_loop())
 
         _LOG.debug("Going Asynchronous")
         try:

--- a/mqtt_io/server.py
+++ b/mqtt_io/server.py
@@ -1147,7 +1147,8 @@ class MqttIo:  # pylint: disable=too-many-instance-attributes
 
     async def _main_loop(self) -> None:
         counter = self.config['mqtt'].get('reconnect_count')
-        while True:
+        reconnect = True
+        while reconnect:
             try:
                 await self._connect_mqtt()
                 # Reset counter
@@ -1175,14 +1176,16 @@ class MqttIo:  # pylint: disable=too-many-instance-attributes
             except asyncio.CancelledError:
                 break
             except MQTTException:
-                if counter == 0:
-                    break
+                reconnect = counter > 0
                 if counter > 0:
                     counter -= 1
                 _LOG.exception('Connection to MQTT-Broker failed')
-                await asyncio.sleep(self.config['mqtt'].get('reconnect_delay'))
+
             finally:
                 self.running.clear()
+                self.mqtt_connected.clear()
+            if reconnect:
+                await asyncio.sleep(self.config['mqtt'].get('reconnect_delay'))
         await self.shutdown()
 
     # Main entry point

--- a/mqtt_io/server.py
+++ b/mqtt_io/server.py
@@ -1147,12 +1147,12 @@ class MqttIo:  # pylint: disable=too-many-instance-attributes
                 self.event_bus.fire(StreamDataSentEvent(stream_conf["name"], data))
 
     async def _main_loop(self) -> None:
-        counter = self.config['mqtt'].get('reconnect-count')
+        counter = self.config['mqtt'].get('reconnect_count')
         while True:
             try:
                 await self._connect_mqtt()
                 # Reset counter
-                counter = self.config['mqtt'].get('reconnect-count')
+                counter = self.config['mqtt'].get('reconnect_count')
                 self.critical_tasks = [
                     self.loop.create_task(coro)
                     for coro in (
@@ -1181,7 +1181,7 @@ class MqttIo:  # pylint: disable=too-many-instance-attributes
                 _LOG.exception('Connection to MQTT-Broker failed')
                 if counter == 0:
                     break
-                await asyncio.sleep(self.config['mqtt'].get('reconnect-delay'))
+                await asyncio.sleep(self.config['mqtt'].get('reconnect_delay'))
             finally:
                 self.running.clear()
         await self.shutdown()

--- a/mqtt_io/server.py
+++ b/mqtt_io/server.py
@@ -19,6 +19,7 @@ from hashlib import sha1
 from importlib import import_module
 from typing import Any, Dict, List, Optional, Tuple, Type, Union, overload
 
+import asyncio_mqtt
 import backoff  # type: ignore
 from typing_extensions import Literal
 
@@ -1171,34 +1172,44 @@ class MqttIo:  # pylint: disable=too-many-instance-attributes
         self._init_stream_modules()
 
         async def run() -> None:
-            try:
-                await self._connect_mqtt()
-
-                self.critical_tasks = [
-                    self.loop.create_task(coro)
-                    for coro in (
-                        self._mqtt_task_loop(),
-                        self._mqtt_rx_loop(),
-                        self._remove_finished_transient_tasks(),
-                    )
-                ]
-
-                self.running.set()
-
-                if self.config["mqtt"].get("ha_discovery", {}).get("enabled"):
-                    self._ha_discovery_announce()
-
+            counter = self.config['mqtt'].get('reconnect-count')
+            while True:
                 try:
-                    await asyncio.gather(*self.critical_tasks)
+                    await self._connect_mqtt()
+                    # Reset counter
+                    counter = self.config['mqtt'].get('reconnect-count')
+                    self.critical_tasks = [
+                        self.loop.create_task(coro)
+                        for coro in (
+                            self._mqtt_task_loop(),
+                            self._mqtt_rx_loop(),
+                            self._remove_finished_transient_tasks(),
+                        )
+                    ]
+
+                    self.running.set()
+
+                    if self.config["mqtt"].get("ha_discovery", {}).get("enabled"):
+                        self._ha_discovery_announce()
+
+                    try:
+                        await asyncio.gather(*self.critical_tasks)
+                    except asyncio.CancelledError:
+                        break
+                    except Exception:  # pylint: disable=broad-except
+                        _LOG.exception("Exception in critical task:")
                 except asyncio.CancelledError:
-                    pass
-                except Exception:  # pylint: disable=broad-except
-                    _LOG.exception("Exception in critical task:")
-            except asyncio.CancelledError:
-                pass
-            finally:
-                self.running.clear()
-                await self.shutdown()
+                    break
+                except asyncio_mqtt.error.MqttError:
+                    if counter > 0:
+                        counter -= 1
+                    _LOG.exception('Connection to MQTT-Broker failed')
+                    if counter == 0:
+                        break
+                    await asyncio.sleep(self.config['mqtt'].get('reconnect-delay'))
+                finally:
+                    self.running.clear()
+            await self.shutdown()
 
         self._main_task = self.loop.create_task(run())
 

--- a/mqtt_io/server.py
+++ b/mqtt_io/server.py
@@ -19,7 +19,7 @@ from hashlib import sha1
 from importlib import import_module
 from typing import Any, Dict, List, Optional, Tuple, Type, Union, overload
 
-import asyncio_mqtt
+import asyncio_mqtt  # type: ignore
 import backoff  # type: ignore
 from typing_extensions import Literal
 

